core/src/main/java/org/apache/accumulo/core/iteratorsImpl/system/SortedMapIterator.java

```diff
@@ -121,4 +143,9 @@ public void init(SortedKeyValueIterator<Key,Value> source, Map<String,String> op
       IteratorEnvironment env) throws IOException {
     throw new UnsupportedOperationException();
   }
+
+  @Override
+  public void setInterruptFlag(AtomicBoolean flag) {
+    this.interruptFlag = flag;
+  }
```

Comment:
Because changing the reference breaks atomicity guarantees, this method looks dubious. It looks like the way this is intended to be used is like a BooleanSupplier (or even more strictly, a one-way latch that can move from false to true, but never back to false). I'm not sure the right solution, but if the interruptible stuff is only on the server-side, it could be cleaned up to avoid using AtomicBoolean and use something more appropriate instead.

Reply:
I agree. I am hesitant to go changing internals with the iterators though since they are so heavily reliant on runtime configuration (one of the reasons it took so long to find this bug in the first place) and I don't think we have adequate test coverage for their behavior.

Reply:
My suggestion would be to use BooleanSupplier in the InterruptibleIterator interface, and call `setInterruptFlag(interruptFlag::get)` (where `interruptFlag` in the caller is of type `AtomicBoolean`). That would remove any confusion over implied atomicity guarantees inside the InterruptibleIterator interface, but wouldn't change any behavior. But, obviously, that would be a follow-on issue. It need not be done as part of this PR.